src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,411 @@
+     * If there are more results, returns a continuation token that can be passed in a subsequent call
+     * to get the next "page" of results.
+     *
+     * Note that folders don't actually exist in S3.
+     * Everything in S3 is an object with a key residing in a bucket.
+     * However, S3 lets you limit results to those residing at a specific "path" specified by delimiters (slashes).
+     * The list of sub-paths is returned in the result set and can be used in subsequent calls.
+     *
+     * A consequence of the fact that folders don't exist is that folders and files are intermingled across all
+     * of the pages.
+     * It's not possible to retrieve an exhaustive list of all folders without traversing all of the pages.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    public async listObjects(request: ListObjectsRequest): Promise<ListObjectsResponse> {
+        const s3 = await this.createS3()
+
+        let response: ListObjectsV2Output
+        try {
+            response = await s3
+                .listObjectsV2({
+                    Bucket: request.bucketName,
+                    Delimiter: DEFAULT_DELIMITER,
+                    MaxKeys: DEFAULT_MAX_KEYS,
+                    Prefix: request.folderPath,
+                    ContinuationToken: request.continuationToken,
+                })
+                .promise()
+        } catch (e) {
+            throw new S3Error(`Failed to list objects for bucket ${request.bucketName}: ${e}`)
+        }
+
+        const files: File[] = _(response.Contents)
+            .reject(file => file.Key === request.folderPath)
+            .map(
+                file =>
+                    new DefaultFile({
+                        key: file.Key!,
+                        partitionId: this.partitionId,
+                        bucketName: request.bucketName,
+                        lastModified: file.LastModified,
+                        sizeBytes: file.Size,
+                    })
+            )
+            .value()
+
+        const folders: Folder[] = _(response.CommonPrefixes)
+            .map(prefix => prefix.Prefix)
+            .compact()
+            .map(path => new DefaultFolder({ path, partitionId: this.partitionId, bucketName: request.bucketName }))
+            .value()
+
+        return {
+            files,
+            folders,
+            continuationToken: response.NextContinuationToken,
+        }
+    }
+
+    /**
+     * Looks up the region for the given bucket
+     *
+     * Note that although there is an S3#GetBucketLocation API,
+     * this is the suggested method of obtaining the region.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    private async lookupRegion(bucketName: string, s3: S3): Promise<string> {
+        try {
+            const response = await s3.headBucket({ Bucket: bucketName }).promise()
+            return response.$response.httpResponse.headers[BUCKET_REGION_HEADER]
```

Comment:
Correct me if I'm wrong, but I'm not seeing a region header?: https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html#API_HeadBucket_ResponseElements

Reply:
Nice catch, it's missing from their docs. I submitted them feedback.